core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2374,10 +2375,17 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
 
         private boolean isMaterializedViewSufficientlyFresh(Session session, QualifiedObjectName name, MaterializedViewDefinition materializedViewDefinition)
         {
-            MaterializedViewFreshness materializedViewFreshness = metadata.getMaterializedViewFreshness(session, name);
+            boolean gracePeriodZero = materializedViewDefinition.getGracePeriod()
+                    .map(Duration::isZero)
+                    .orElse(false);
+
+            // Do not allow connectors to optimize freshness checks when the grace period is zero,
+            // so that this special case is handled in the analyzer rather than duplicating the logic across all connectors.
+            boolean considerGracePeriod = !gracePeriodZero;
```

Comment:
I am not convinced this is necessary.

i know it follows my own logic
https://github.com/trinodb/trino/blob/3f8b4e056227f17be8f2a410877407c950250b37/core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java#L2397-L2398

but it makes code flow more complicated.

anyway, no change requested